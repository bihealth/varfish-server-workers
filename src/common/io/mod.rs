//! Common, IO-related code.

pub mod std;
