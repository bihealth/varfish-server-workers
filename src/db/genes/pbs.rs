//! Data structures for (de-)serialization as generated by `prost-build`.

include!(concat!(env!("OUT_DIR"), "/varfish.v1.gene.rs"));
