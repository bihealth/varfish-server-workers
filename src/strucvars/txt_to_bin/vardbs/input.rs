//! Code supporting the I/O of public database records and a common
//! `InputRecord` for common representation.

use serde::Deserialize;
use tracing::error;

use crate::strucvars::aggregate::output::Record as InhouseDbRecord;
use crate::strucvars::query::schema::SvType;

/// dbVar database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DbVarRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// number of overall carriers
    pub num_carriers: u32,
    /// type of the SV
    pub sv_type: String,
}

/// DGV database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DgvRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural variant type
    sv_type: String,
    /// Number of observed gains.
    observed_gains: u32,
    /// Number of observed losses
    observed_losses: u32,
}

/// DGV gold standard database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DgvGsRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin_outer: i32,
    /// outer end position, 0-based
    pub end_outer: i32,
    /// The structural variant type
    pub sv_sub_type: String,
    /// Number of carriers.
    pub num_carriers: u32,
}

/// ExAC CNV database record as read from TSV file for deserialization
/// from TSV.
#[derive(Deserialize, Debug)]
pub struct ExacRecord {
    /// chromosome name
    pub chromosome: String,
    /// outer start position, 0-based
    pub begin: i32,
    /// outer end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub sv_type: String,
}

/// Thousand Genomes SV database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct G1kRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub sv_type: String,
    /// Number of hom. alt. alleles.
    pub n_homalt: u32,
    /// Number of het. alt. alleles.
    pub n_het: u32,
}

/// gnomAD SV v2 database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadSv2Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of homozygous alternative carriers
    pub n_homalt: u32,
    /// Number of heterozygous carriers
    pub n_het: u32,
}

/// gnomAD SV v4 database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadSv4Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of male homozygous reference allele carriers.
    pub male_n_homref: u32,
    /// Number of male heterozygous alternate allele carriers.
    pub male_n_het: u32,
    /// Number of male homozygous alternate allele carriers.
    pub male_n_homalt: u32,
    /// Number of male hemizygous alternate allele carriers.
    pub male_n_hemiref: u32,
    /// Number of male hemizygous reference allele carriers.
    pub male_n_hemialt: u32,
    /// Number of female homozygous reference allele carriers.
    pub female_n_homref: u32,
    /// Number of female heterozygous alternate allele carriers.
    pub female_n_het: u32,
    /// Number of female homozygous alternate allele carriers.
    pub female_n_homalt: u32,
    /// Number of samples at this site (CNV only).
    pub cnv_n_total: u32,
    /// Number of samples with a CNV at this site (CNV only).
    pub cnv_n_var: u32,
}

/// gnomAD CNV v$ database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadCnv4Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of samples at this site (passing QC).
    pub n_total: u32,
    /// Number of samples with a CNV at this site (passing QC).
    pub n_var: u32,
}

/// Common type to convert input data to.
pub struct InputRecord {
    /// Chromosome of start position.
    pub chromosome: String,
    /// Chromosome of end position.
    pub chromosome2: String,
    /// SV type
    pub sv_type: SvType,
    /// 0-based begin position
    pub begin: i32,
    /// 0-based end position
    pub end: i32,
    /// Number of carriers (or alleles), depending on database.
    pub count: u32,
}

impl TryInto<Option<InputRecord>> for InhouseDbRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome,
            chromosome2: self.chromosome2,
            sv_type: self.sv_type,
            begin: self.begin,
            end: self.end,
            count: self.carriers,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DbVarRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.split(';').next().unwrap() {
            "alu_insertion"
            | "herv_insertion"
            | "insertion"
            | "line1_insertion"
            | "mobile_element_insertion"
            | "novel_sequence_insertion"
            | "sva_insertion" => SvType::Ins,
            "copy_number_gain" | "duplication" | "tandem_duplication" => SvType::Dup,
            "alu_deletion" | "copy_number_loss" | "deletion" | "herv_deletion"
            | "line1_deletion" | "sva_deletion" => SvType::Del,
            "copy_number_variation" => SvType::Cnv,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: 1,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DgvRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_ref() {
            "alu deletion"
            | "deletion"
            | "herv deletion"
            | "line1 deletion"
            | "mobile element deletion"
            | "loss"
            | "sva deletion" => SvType::Del,
            "alu insertion"
            | "herv insertion"
            | "insertion"
            | "line1 insertion"
            | "mobile element insertion"
            | "novel sequence insertion"
            | "sva insertion" => SvType::Ins,
            "duplication" | "gain" | "tandem duplication" => SvType::Dup,
            "sequence alteration" | "complex" => return Ok(None), // skip
            "gain+loss" | "CNV" => SvType::Cnv,
            "inversion" => SvType::Inv,
            "OTHER" => return Ok(None), // skip
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: self.observed_gains + self.observed_losses,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DgvGsRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_sub_type.as_ref() {
            "Gain" => SvType::Dup,
            "Loss" => SvType::Del,
            _ => {
                error!("sv_type = {}", &self.sv_sub_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin_outer,
            end: self.end_outer,
            sv_type,
            count: self.num_carriers,
        }))
    }
}

impl TryInto<Option<InputRecord>> for ExacRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_ref() {
            "DUP" => SvType::Dup,
            "DEL" => SvType::Del,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: 1,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadSv2Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.svtype.as_str() {
            "CPX" => return Ok(None), // no correspondence
            "CTX" | "BND" => SvType::Bnd,
            "DEL" => SvType::Del,
            "DUP" => SvType::Dup,
            "INS" => SvType::Ins,
            "INV" => SvType::Inv,
            "MCNV" => SvType::Cnv,
            _ => {
                error!("sv_type = {}", &self.svtype);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin - 1,
            end: self.end,
            sv_type,
            count: self.n_homalt + self.n_het,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadCnv4Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type: match self.svtype.as_str() {
                "DEL" => SvType::Del,
                "DUP" => SvType::Dup,
                _ => {
                    error!("sv_type = {}", &self.svtype);
                    return Err("unknown SV type");
                }
            },
            count: self.n_var,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadSv4Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type: match self.svtype.as_str() {
                "BND" => SvType::Bnd,
                "CNV" => SvType::Cnv,
                "DEL" => SvType::Del,
                "DUP" => SvType::Dup,
                "INS" => SvType::Ins,
                "INV" => SvType::Inv,
                _ => {
                    error!("sv_type = {}", &self.svtype);
                    return Err("unknown SV type");
                }
            },
            count: self.male_n_het
                + self.male_n_homalt
                + self.male_n_hemialt
                + self.female_n_het
                + self.female_n_homalt
                + self.cnv_n_var,
        }))
    }
}

impl TryInto<Option<InputRecord>> for G1kRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_str() {
            "CN0" | "CNV" => SvType::Cnv,
            "DEL" => SvType::Del,
            "DEL_ALU" | "DEL_HERV" | "DEL_LINE1" | "DEL_SVA" => SvType::Del,
            "DUP" => SvType::Dup,
            "INV" => SvType::Inv,
            "INS" | "INS:ME:ALU" | "INS:ME:LINE1" | "INS:ME:SVA" => SvType::Ins,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: self.n_homalt + self.n_het,
        }))
    }
}
