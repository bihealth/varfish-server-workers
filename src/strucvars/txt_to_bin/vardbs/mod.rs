//! Code for converting other structural variant database to binary (incl. in-house).

use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::time::Instant;

use anyhow::anyhow;
use prost::Message;
use thousands::Separable;

use crate::common::{build_chrom_map, trace_rss_now};
use crate::pbs::svs::{BackgroundDatabase, BgDbRecord};
use crate::strucvars::aggregate::output::Record as InhouseDbRecord;
use crate::strucvars::query::schema::SvType;

use self::input::InputRecord;

mod input;

/// Known input file types.
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum InputFileType {
    Dbvar,
    Dgv,
    DgvGs,
    Exac,
    G1k,
    GnomadSv2,
    GnomadCnv4,
    GnomadSv4,
    InhouseDb,
}
/// Deserialize from CSV reader to an `Option<records::InputRecord>`
fn deserialize_loop<Rec>(
    reader: &mut csv::Reader<Box<dyn std::io::BufRead>>,
) -> Result<Vec<BgDbRecord>, anyhow::Error>
where
    Rec: core::fmt::Debug + TryInto<Option<InputRecord>> + for<'de> serde::Deserialize<'de>,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: core::fmt::Debug,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: Send,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: std::marker::Sync,
{
    let chrom_map = build_chrom_map();
    let mut result = Vec::new();

    for record in reader.deserialize() {
        let record: Rec = record?;
        let maybe_record: Option<InputRecord> = record
            .try_into()
            .map_err(|err| anyhow!("problem with parsing: {:?}", &err))?;
        if let Some(record) = maybe_record {
            result.push(BgDbRecord {
                chrom_no: *chrom_map
                    .get(&record.chromosome)
                    .unwrap_or_else(|| panic!("unknown chrom: {:?}", &record.chromosome))
                    as i32,
                chrom_no2: *chrom_map
                    .get(&record.chromosome2)
                    .unwrap_or_else(|| panic!("unknown chrom2: {:?}", &record.chromosome2))
                    as i32,
                sv_type: match record.sv_type {
                    SvType::Del => crate::pbs::svs::SvType::Del,
                    SvType::Dup => crate::pbs::svs::SvType::Dup,
                    SvType::Inv => crate::pbs::svs::SvType::Inv,
                    SvType::Ins => crate::pbs::svs::SvType::Ins,
                    SvType::Bnd => crate::pbs::svs::SvType::Bnd,
                    SvType::Cnv => crate::pbs::svs::SvType::Cnv,
                } as i32,
                start: record.begin + 1,
                stop: record.end,
                count: record.count,
            });
        }
    }

    Ok(result)
}

/// Branch around `deserialize_loop`.
pub fn deserialize_branch(
    input_type: InputFileType,
    reader: &mut csv::Reader<Box<dyn std::io::BufRead>>,
) -> Result<Vec<BgDbRecord>, anyhow::Error> {
    match input_type {
        InputFileType::Dbvar => deserialize_loop::<input::DbVarRecord>(reader),
        InputFileType::Dgv => deserialize_loop::<input::DgvRecord>(reader),
        InputFileType::DgvGs => deserialize_loop::<input::DgvGsRecord>(reader),
        InputFileType::Exac => deserialize_loop::<input::ExacRecord>(reader),
        InputFileType::G1k => deserialize_loop::<input::G1kRecord>(reader),
        InputFileType::InhouseDb => deserialize_loop::<InhouseDbRecord>(reader),
        InputFileType::GnomadSv2 => deserialize_loop::<input::GnomadSv2Record>(reader),
        InputFileType::GnomadCnv4 => deserialize_loop::<input::GnomadCnv4Record>(reader),
        InputFileType::GnomadSv4 => deserialize_loop::<input::GnomadSv4Record>(reader),
    }
}

/// Perform conversion to protobuf `.bin` file.
pub fn convert_to_bin<P, Q>(
    path_input_tsv: P,
    path_output: Q,
    input_type: InputFileType,
) -> Result<(), anyhow::Error>
where
    P: AsRef<Path>,
    Q: AsRef<Path>,
{
    // Setup CSV reader for BED file - header is written as comment and must be
    // ignored.
    let mut reader = csv::ReaderBuilder::new()
        .has_headers(false)
        .comment(Some(b'#'))
        .delimiter(b'\t')
        .from_reader(mehari::common::io::std::open_read_maybe_gz(
            path_input_tsv.as_ref(),
        )?);
    let before_parsing = Instant::now();

    let records = deserialize_branch(input_type, &mut reader)?;
    let bg_db = BackgroundDatabase { records };

    tracing::debug!(
        "total time spent reading {} records: {:?}",
        bg_db.records.len().separate_with_commas(),
        before_parsing.elapsed()
    );
    trace_rss_now();

    let before_writing = Instant::now();
    let mut output_file = File::create(&path_output)?;
    output_file.write_all(&bg_db.encode_to_vec())?;
    output_file.sync_all()?;
    tracing::debug!(
        "total time spent writing {} records: {:?}",
        bg_db.records.len().separate_with_commas(),
        before_writing.elapsed()
    );

    Ok(())
}

#[cfg(test)]
mod test {
    use super::InputFileType;

    #[rstest::rstest]
    #[case::dbvar(
        InputFileType::Dbvar,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dbvar.bed.gz"
    )]
    #[case::dgv(
        InputFileType::Dgv,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dgv.bed.gz"
    )]
    #[case::dgv_gs(
        InputFileType::DgvGs,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dgv_gs.bed.gz"
    )]
    #[case::exac(
        InputFileType::Exac,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/exac.bed.gz"
    )]
    #[case::g1k(
        InputFileType::G1k,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/g1k.bed.gz"
    )]
    #[case::gnomad_sv2(
        InputFileType::GnomadSv2,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/gnomad_sv.bed.gz"
    )]
    #[case::gnomad_cnv4(
        InputFileType::GnomadCnv4,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch38/strucvar/gnomad-cnv.bed.gz"
    )]
    #[case::gnomad_sv4(
        InputFileType::GnomadSv4,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch38/strucvar/gnomad-sv.bed.gz"
    )]
    #[case::inhouse_db(
        InputFileType::InhouseDb,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/inhouse.tsv"
    )]
    fn test_deserialize_branch(
        #[case] input_type: InputFileType,
        #[case] path_input: &str,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!(
            "{:?}-{}",
            input_type,
            path_input
                .split('/')
                .last()
                .unwrap()
                .split('.')
                .next()
                .unwrap()
        );

        let mut reader = csv::ReaderBuilder::new()
            .has_headers(false)
            .comment(Some(b'#'))
            .delimiter(b'\t')
            .from_reader(mehari::common::io::std::open_read_maybe_gz(path_input)?);

        let records = super::deserialize_branch(input_type, &mut reader)?;
        insta::assert_yaml_snapshot!(records);

        Ok(())
    }
}
