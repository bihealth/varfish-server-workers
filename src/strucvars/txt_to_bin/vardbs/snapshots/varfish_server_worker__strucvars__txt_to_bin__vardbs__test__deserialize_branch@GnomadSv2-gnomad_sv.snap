---
source: src/strucvars/txt_to_bin/vardbs/mod.rs
expression: records
---
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 10641
  stop: 10643
  count: 140
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 20999
  stop: 26000
  count: 138
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 39999
  stop: 47200
  count: 852
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 50940
  stop: 51053
  count: 81
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 54664
  stop: 54666
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 54719
  stop: 54721
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 62398
  stop: 62489
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 66349
  stop: 66427
  count: 12
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 66530
  stop: 66532
  count: 5
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 66575
  stop: 66577
  count: 5
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 79209
  stop: 79211
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 86748
  stop: 86750
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 121034
  stop: 121130
  count: 129
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 136999
  stop: 155500
  count: 2156
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 141474
  stop: 155000
  count: 127
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 149999
  stop: 155000
  count: 1075
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 151999
  stop: 177400
  count: 72
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 156999
  stop: 166000
  count: 380
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 167899
  stop: 174000
  count: 76
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 243915
  stop: 244232
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 251747
  stop: 251749
  count: 7
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 317718
  stop: 367000
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 319999
  stop: 327650
  count: 3283
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 353799
  stop: 363550
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 353999
  stop: 359000
  count: 5225
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 360999
  stop: 366100
  count: 22
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 363999
  stop: 401750
  count: 4087
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 364299
  stop: 370000
  count: 659
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 366999
  stop: 384100
  count: 1978
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 368999
  stop: 374000
  count: 101
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 371199
  stop: 376900
  count: 332
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 375999
  stop: 382450
  count: 484
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 380499
  stop: 386800
  count: 48
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 383999
  stop: 390000
  count: 283
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 395499
  stop: 471184
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 406349
  stop: 415250
  count: 1536
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 412999
  stop: 461500
  count: 40
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 433999
  stop: 441000
  count: 1560
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 439999
  stop: 471050
  count: 13
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 445999
  stop: 461000
  count: 695
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 455999
  stop: 468000
  count: 472
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 459499
  stop: 467000
  count: 26
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 460999
  stop: 466000
  count: 349
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 461999
  stop: 470342
  count: 1251
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 462999
  stop: 468000
  count: 167
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 521999
  stop: 638000
  count: 618
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 530885
  stop: 530887
  count: 13
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 530885
  stop: 530887
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 531072
  stop: 531074
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 531099
  stop: 613750
  count: 6666
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 531999
  stop: 537000
  count: 1330
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 535816
  stop: 536117
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 535999
  stop: 541000
  count: 1055
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 536037
  stop: 536215
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 544839
  stop: 544841
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 547598
  stop: 547802
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 548449
  stop: 564400
  count: 872
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 569999
  stop: 580000
  count: 698
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 569999
  stop: 593000
  count: 502
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 571749
  stop: 585025
  count: 3523
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 574749
  stop: 581950
  count: 534
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 582699
  stop: 593000
  count: 600
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 591099
  stop: 596200
  count: 366
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 598599
  stop: 605000
  count: 1028
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 603329
  stop: 603403
  count: 629
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 604999
  stop: 613000
  count: 3
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 656999
  stop: 665000
  count: 54
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 667724
  stop: 724000
  count: 125
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 668629
  stop: 668631
  count: 189
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 668629
  stop: 850205
  count: 220
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 685999
  stop: 700425
  count: 3
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 711672
  stop: 711674
  count: 21
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 713366
  stop: 713368
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 718807
  stop: 718809
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 3
  start: 720319
  stop: 720321
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 721937
  stop: 721939
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 721937
  stop: 773013
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 727999
  stop: 733650
  count: 680
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 728091
  stop: 728093
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 728091
  stop: 776880
  count: 5
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 737016
  stop: 737018
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 738609
  stop: 742016
  count: 12
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 744699
  stop: 751000
  count: 274
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 746763
  stop: 747331
  count: 35
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 750054
  stop: 750056
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 754259
  stop: 857641
  count: 97
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 755402
  stop: 755404
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 755451
  stop: 756062
  count: 35
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 755459
  stop: 755461
  count: 1
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 755633
  stop: 755635
  count: 709
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 755706
  stop: 755892
  count: 816
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 755890
  stop: 755892
  count: 2
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 755891
  stop: 758321
  count: 671
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 756038
  stop: 756040
  count: 467
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 756258
  stop: 758459
  count: 4180
- chrom_no: 0
  chrom_no2: 0
  sv_type: 4
  start: 756376
  stop: 756378
  count: 49
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 756571
  stop: 758603
  count: 38
- chrom_no: 0
  chrom_no2: 0
  sv_type: 0
  start: 757420
  stop: 834328
  count: 7
- chrom_no: 0
  chrom_no2: 0
  sv_type: 1
  start: 757478
  stop: 758945
  count: 3
